src/lib/utils.cc

```diff
@@ -682,4 +724,247 @@ TEST_CASE("get_nuc") {
 }
 // GCOVR_EXCL_STOP
 
+/**
+ * @brief Reorder pair of input sequences so that reference is at position zero.
+ *
+ * @param[in,out] aln coati::alignment_t alignment data.
+ */
+void order_ref(coati::alignment_t& aln) {
+    if(aln.data.names[0] == aln.refs) {
+        // already the first sequence: do nothing
+    } else if(aln.data.names[1] == aln.refs || aln.rev) {  // swap sequences
+        std::swap(aln.data.names[0], aln.data.names[1]);
+        std::swap(aln.data.seqs[0], aln.data.seqs[1]);
+        if(aln.data.fsts.size() > 0) {
+            std::swap(aln.data.fsts[0], aln.data.fsts[1]);
+        }
+    } else {  // aln.refs was specified and doesn't match any seq names
+        throw std::invalid_argument("Name of reference sequence not found.");
+    }
+}
+
+/**
+ * @brief Read and validate input sequences.
+ *
+ * @param[in,out] aln coati::alignment_t input sequences and alignment info.
+ *
+ */
+void process_marginal(coati::alignment_t& aln) {
+    // different functions for alignpair/msa/fst/.../ ?
+
+    if(aln.data.size() != 2) {
+        throw std::invalid_argument("Exactly two sequences required.");
+    }
+
+    // set reference sequence as first sequence
+    if(!aln.refs.empty() || aln.rev) {
+        order_ref(aln);
+    }
+
+    // check that length of ref is multiple of 3 and gap unit length
+    size_t len_a = aln.seq(0).length();
+    size_t len_b = aln.seq(1).length();
+    if(len_a % 3 != 0 || len_a % aln.gap.len != 0) {
+        throw std::invalid_argument(
+            "Length of reference sequence must be multiple of 3 and gap unit "
+            "length.");
+    }
+
+    // check that length of descendant is multiple of gap unit length
+    if(len_b % aln.gap.len != 0) {
+        throw std::invalid_argument(
+            "Length of descendant sequence must be multiple of gap unit "
+            "length.");
+    }
+
+    // handle ending stop codons
+    trim_end_stops(aln.data);
+}
+
+/**
+ * @brief Trim end stop codons.
+ *
+ * @param[in,out] aln coati::alignment_t input sequences and alignment info.
+ */
+void trim_end_stops(coati::data_t& data) {
+    for(size_t i = 0; i < data.size(); ++i) {
+        std::string seq{data.seqs[i]};
+        size_t len = seq.length();
+        std::string last_cod{seq.substr(len - 3)};
```

Comment:
You should be able to use a string_view here.

Reply:
Done, thanks.